### DISCUSSION TITLE
Keybinding quickstart

### DISCUSSION THREAD

**Discussion by mavit** [kickingvegas/casual#99] [Ideas]
Currently, each of the Casual transients needs to be bound to a key individually.  This requires a fair bit of looking through the documentation, then copying-and-pasting to the user’s init file.

Perhaps there could be a convenience function that goes ahead and sets the bindings suggested in the documentation?  I’m imagining something along the lines of the following:

``` emacs-lisp
(defcustom casual-key "C-o"
  "The key that `casual-keymap-set' will bind to open Casual's transient menus."
  :type 'key
  :group 'casual
  :package-version '("casual" . "2.3.0"))

(defun casual-keymap-set (&optional consistent)
  "Add keymappings to open Casual transient menus.

If CONSISTENT is true, also alter mode keymaps so that their keys match
those used in the Casual transient menus."

  (interactive)

  (with-eval-after-load 'org-agenda
    (keymap-set org-agenda-mode-map casual-key #'casual-agenda-tmenu)
    (when consistent
      (keymap-set org-agenda-mode-map "M-j" #'org-agenda-clock-goto)
      (keymap-set org-agenda-mode-map "J" #'bookmark-jump)))

  ...)

```

Alternatively, perhaps a global minor mode would be a better design.

**Comment by kickingvegas** [comment]
@mavit Thanks much for sharing this. Minimizing the work required to install and configure Casual is of great interest to me. This however has turned out to be much more challenging than expected as I have struggled with making an implementation that addresses the following concerns:

-   The implementation needs to work with late binding/lazy loading so that packages are not loaded until necessary.
    -   Working with late binding is especially of value to `use-package`.
-   The implementation needs to allow for user preference in key bindings.

Early work on this has been done in #26 where an alist is used as a lookup table to invoke the appropriate Transient menu when using a global key-binding. My implementation worked for some modes but not for others (Dired for instance). So I’ve not pushed it out.

It's unclear to me that the implementation example you've provided will byte-compile cleanly unless I first invoke `require` on the packages whose symbol is referenced. Doing so however will defeat lazy loading.

Perhaps others may have better insights into this as my Elisp skills are relatively novice.

**Reply by mavit** [reply]
The above byte-compiles without warnings for me provided that I:

- First `(eval-when-compile (require 'org-agenda))`.
- Reference `'org-agenda-clock-goto` rather than `#'org-agenda-clock-goto`.

However, I think I prefer the dispatcher-based approach discussed in #26.
